Add argument asserts to copyForwards and copyBackwards

moosichu opened the pull request:
Helps document in code what the comments already say. I had to fixup a comptime use case of these functions in Encoding.zig (as you can't retrieve the pointer value at comptime...). However, @memcpy seems to work at comptime now so it resolves a pre-existing TODO there anyway!

It does raise the question of how you should handle comptime memcpys with overlapping arrays though - as ideally you would be able to still validate the ptr ordering in that case as well?

This also fixes a bug where an assert in `copyBackwards` was effectively doing nothing due to an `@setRuntimeSafety` call in the same scope.

moosichu commented:
I mainly made this because I got the intended uses the wrong way around 🤦

moosichu commented:
Obviously I completed a TODO but didn't remove it 🤦  (that is now fixed!)

rohlem commented:
You can put this check behind an `if(!@inComptime())` to make it runtime-only. That way `comptime` calls won't be affected.

There is another issue however in that two non-overlapping slices (say from disconnected allocations) should also be free to use either version (even if `@memcpy` would be more optimal for that case - maybe the same code path sometimes overlaps in the correct direction and sometimes doesn't overlap at all).
So the correct check IMO would be to only assert the ordering if the slices actually do overlap.

<details><summary>code simplified from my own codebase for reference:</summary>
<p>

```zig
fn copyToBeforeSelfOrNoOverlap(self_source: anytype, destination: anytype) void {
    if (std.debug.runtime_safety and !@inComptime()) {
        const source_begin_address = @intFromPointer(self_source.ptr);
        const source_end_address = source_begin_address + self_source.len;
        const destination_begin_address = @intFromPointer(destination.ptr);
        if ((destination_begin_address > source_begin_address) and (destination_begin_address < source_end_address)) {
            panic("invalid use of copyToBeforeSelfOrNoOverlap detected: destination lies after source; should have used copyToAfterSelfOrNoOverlap!\ntried copying from {} to {}", .{ self_source, destination });
        }
    }
    std.mem.copyForwards(Element, destination, self_source); //old version before this PR
}
fn copyToAfterSelfOrNoOverlap(self_source: anytype, destination: anytype) void {
    if (std.debug.runtime_safety and !@inComptime()) {
        const source_begin_address = @intFromPtr(self_source.ptr);
        const destination_begin_address = @intFromPtr(destination.ptr);
        const destination_end_address = destination_begin_address + destination.len;
        const destination_end_used_address = destination_begin_address + self_source.len;
        if ((source_begin_address > destination_begin_address) and (source_begin_address < destination_end_address)) {
            if (source_begin_address > destination_end_used_address) {
                std.log.warn("potentially invalid use of copyToAfterSelfOrNoOverlap detected: destination lies before source (but dynamically ok because of source length); should probably have used copyToBeforeSelfOrNoOverlap!\ncopying from {} to {}", .{ self_source, destination });
                std.debug.dumpCurrentStackTrace(null);
            } else {
                panic("invalid use of copyToAfterSelfOrNoOverlap detected: destination lies before source; should have used copyToBeforeSelfOrNoOverlap!\ntried copying from {} to {}", .{ self_source, destination });
            }
        }
    }
    std.mem.copyBackwards(Element, destination, self_source);
}
```
</p>
</details>

moosichu commented:
I'm just testing that more comprehensive solution now...

However - something I haven't added to the PR but easily could is this as well using the building blocks this PR now contains:
```zig
/// Copy all or source into dest at position 0.
/// dest.len must be >= source.len.
/// Deals with overlapping/aliasing slices forwards or backwards.
/// Note: Cannot be called at comptime as there is currently no mechanism for
/// determining how comptime slices overlap.
pub fn copySafe(comptime T: type, dest: []T, source: []const T) void {
    if (@inComptime()) {
        @compileError("copySafe cannot currently be called at comptime, as there is no mechanism for determining if two slices alias each other available");
    }
    assert(dest.len >= source.len);
    if (!slicesOverlap(T, dest, source)) {
        @memcpy(dest[0..source.len], source);
        return;
    }
    if (@intFromPtr(source.ptr) >= @intFromPtr(dest.ptr)) {
        copyForwards(T, dest, source);
    } else {
        copyBackwards(T, dest, source);
    }
}
```

rohlem commented:
@moosichu regarding `copySafe`, so far I don't have any callsites in my own code that would benefit from this mechanism.
That said, it's possible legitimate use cases exist, also it might be an easier option for newcomers / less performance-oriented code.
I'd vote for having it (should do no harm), but ultimately up to the maintainers to decide.

Regarding the implementation, I'm not sure about the branch deciding between `@memcpy` and `copyForward`:
 - My gut feeling says that for plain `copyForwards` and `copyBackwards` the compiler "should be smart enough" to understand the pattern, meaning an extra branch to plain `@memcpy` shouldn't improve performance.
 - If that is true, it probably won't help performance in `copySafe` either.

But that's also a decision for maintainers to weigh in, maybe some backends would benefit from a builtin `@safeMemcpy`?
Either way, having a concrete/realistic usage scenario as motivation would probably help the discussion.

rohlem commented on lib/std/mem.zig:
```diff
@@ -196,22 +196,54 @@ test "Allocator.resize" {
+    const a_end = a_begin + (@sizeOf(T) * slice_a.len);
+    const b_end = b_begin + (@sizeOf(T) * slice_b.len);
+
+    if (a_begin > b_begin) {
+        // a starts after b, so confirm b_end doesn't overlap a_begin!
+        return b_end <= a_begin;
+    } else if (a_begin < b_begin) {
+        // b starts after a, so confirm a_end doesn't overlap b_begin!
+        return a_end <= b_begin;
+    }
```
The new `copyForwards`, `copyBackwards` enforce pointer ordering if we return `true` here.
Since there's only danger of overwriting if `a_end != b_begin`, I think the `return`-ed condition should be `<` instead of `<=` - I think that's the source of the current CI errors.

moosichu replied:
yeah! independetntly noticed that when reviewing things myself, cheers for that! 👍

moosichu commented:
Updated the PR to fix some logic and clarify what is going on in the implementation.

rohlem commented:
I realized there's still optimization potential:
Because each call site really only cares about one-directional overlap, instead of a bi-directional `sliceOverlap` check,
a more optimal implementation would checking `if (sliceContainsPointer(a, b.ptr)) unreachable;`.
That effectively reduces the number of if-s on the hot path from 2 to 1.

moosichu commented:
yeah... I considered something like that... but was really struggling to come up with a good, descriptive name for it.

That is a really neat one though so I will put something together tomorrow and check that in!

andrewrk commented on lib/std/mem.zig:
```diff
@@ -196,26 +196,70 @@ test "Allocator.resize" {
 /// dest.len must be >= source.len.
 /// If the slices overlap, dest.ptr must be <= src.ptr.
 pub fn copyForwards(comptime T: type, dest: []T, source: []const T) void {
+    assert(dest.len >= source.len);
```
```suggestion
```

This is already asserted by the for loop below.

andrewrk commented on src/arch/x86_64/Encoding.zig:
```diff
@@ -793,10 +793,8 @@ const mnemonic_to_encodings_map = init: {
-        // compile error for this copy happening at compile time.
-        std.mem.copyForwards(Op, &data.ops, entry[2]);
-        std.mem.copyForwards(u8, &data.opc, entry[3]);
+        @memcpy(data.ops[0..entry[2].len], entry[2]);
+        @memcpy(data.opc[0..entry[3].len], entry[3]);
```
Thank you for fixing these :+1:

andrewrk commented on lib/std/mem.zig:
```diff
@@ -196,26 +196,70 @@ test "Allocator.resize" {
 
+/// Returns true if the slices overlap
+/// This is an *exclusive* range track
+pub fn slicesOverlap(comptime T: type, slice_a: []const T, slice_b: []const T) bool {
```
Everything in std.mem already has to do with slices:

```suggestion
pub fn alias(comptime T: type, slice_a: []const T, slice_b: []const T) bool {
```

andrewrk commented on lib/std/mem.zig:
```diff
@@ -196,26 +196,70 @@ test "Allocator.resize" {
+    const a_end = @intFromPtr(slice_a.ptr + slice_a.len);
+    const b_end = @intFromPtr(slice_b.ptr + slice_b.len);
+
+    // This coment is written assuming a_begin is > b_begin... the other case is
+    // an exercise for the reader ;)
+    //
+    // The first element of a comes after slice_b, so we need to confirm that
+    // the last element of slice_b doesn't come at or after the first element of
+    // slice_a.
+    // b_end points to 1 element beyond the range of slice_b, so we need to
+    // check that b_end > a_begin - there is no aliasing if b_end == a_begin
+    // because b_end isn't actually part of b_slice itself!
+    if (a_begin > b_begin) {
+        return b_end > a_begin;
+    } else if (b_begin > a_begin) {
+        return a_end > b_begin;
+    }
+    assert(a_begin == b_begin);
+    return true;
```
I find all of these comments and logic confusing, when you could do simply instead:

```suggestion
    return !(a_begin >= b_end or b_begin >= a_end);
```

moosichu replied:
🙈  I can overcomplicate simple things sometimes! That makes so much more sense!

andrewrk reviewed (changes requested):
In addition to the problems identified in this review, the entire idea of checking whether aliasing is occurring in userland is flawed. You already noticed that it cannot be done at compile time, but there is more. This would wreck pointer provenance-based optimizations in ReleaseSafe mode because of the `@intFromPtr` calls.

I think general-purpose alias checking needs to be done via a language builtin instead.

moosichu commented:
> I realized there's still optimization potential: Because each call site really only cares about one-directional overlap, instead of a bi-directional `slicesOverlap` check, a more optimal implementation would be checking `if (sliceContains(a, b.ptr)) unreachable;`. That effectively reduces the number of if-s on the hot path from 2 to 1.

I've decided not to do this... based on Andrew's comments if an @alias builtin will be introduced it makes sense to have a user-space implementation of that for now! (as that is what will have to be used once the builtin is introduced).

> In addition to the problems identified in this review, the entire idea of checking whether aliasing is occurring in userland is flawed. You already noticed that it cannot be done at compile time, but there is more. This would wreck pointer provenance-based optimizations in ReleaseSafe mode because of the `@intFromPtr` calls.
> 
> I think general-purpose alias checking needs to be done via a language builtin instead.

Based on this I hope you don't mind that I've gone ahead and added #17613. I can leave this PR open for now.

I personally think this might still be worth merging in the short term for the safety benefits it would bring (even if not ideal) and then fix it once the builtin is introduced. But otherwise it can be left open and modified once someone has done #17613. (I want to focus on the archiver for now... so won't touch it in the short-term at least).

rohlem commented:
> You already noticed that it cannot be done at compile time, but there is more.

I do think that checking whether one pointer points within a still-modifiable slice of a `comptime` memory isle should be implementable as a compiler builtin.
With `comptime` pointers-to-mutable, the compiler already has to allow different pointers (with the same value) to modify the same location, so it's already required to track identity.
That is exactly what this use case would require, every scenario outside of that can gracefully result in a compile error.

I also think that the name `alias` is just as fitting for arguments `(slice, pointer)` as it is for arguments `(slice, slice)`.
The second argument can be made `anytype` and support both use cases (`anytype` parameters similarly exist in other `std.mem` functions as well).

Just my thoughts; feel free to let me know if I'm wrong / you disagree with any of the above course.
Otherwise I'll probably try implementing it myself (though not in the immediate future, more likely in a year or so).

moosichu commented:
> I also think that the name alias is just as fitting for arguments (slice, pointer) as it is for arguments (slice, slice).
The second argument can be made anytype and support both use cases (anytype parameters similarly exist in other std.mem functions as well).

I've implemented this in a follow-up commit. Very happy to revert it based on feedback also, but thought it would be worth having for comparison's sake.

moosichu commented:
No! You aren't being pedantic at all. Rushed this a little this morning before heading out and forgot about that bit so will do that later.

moosichu commented:
saw the tests fail.. theres obviously some failure of logic but need to head to bed now and will look properly tomorrow!

rohlem commented:
Well, I'm embarassed, the CI errors trigger because the new formulation doesn't work for a no-op copy where `source.ptr == destination.ptr`. Sorry for that oversight.
The optimal condition should be `ptr_int < range_end and ptr_int > range_begin`, but I'm not sure what a good name for the function would then be. `aliasDifferent`, `aliasAtOffset`?
We could also write the asserted condition as `!alias(...) or (source.ptr == dest.ptr)` and hope the optimizer merges them after inlining (though it probably wouldn't in `Debug` builds).

moosichu commented:
I've gone for the:
`!alias(...) or (source.ptr == dest.ptr)`
approach without inlining, mainly because of the presumption is that `@alias` will eventually be a builtin with the API I've implemented, so that will be the building block the function will eventually have to use anyway.
